### ISSUE TITLE
RIKEN: Solenoid PSU

### ISSUE THREAD

**Issue by kjwoodsISIS** [ISISComputingGroup/IBEX#5431]
As a beamline engineer, I need an interface to the new PSU for the RIKEN solenoid, so that I can control the current being supplied to the solenoid.

#### Acceptance Criteria
1. I can control the PSU via the IBEX GUI.
1. I can control the PSU via genie_python scripts.

#### Notes
1. The RIKEN solenoid is an existing device on the RIKEN beamline.  It is an old system and is due to be replaced within the next 2-3 years.
   * Because it is an old system, it has an equally old power supply.  There is a risk that the old power supply might fail before a new solenoid can be installed.
   * The old power supply is not controlled via IBEX, nor is there any intention that it should be.
   * The beamline engineers would like to have a spare PSU that they can have ready, "sitting-on-the-shelf" to replace the old PSU.
1. The spare PSU will be a TDK Lambda Genesys power supply.
1. The PSU itself is expected to be directly controlled by a PLC (details to follow). IBEX wilol communicate to the PLC.  The PLC is expected to control:
   1. ramping of the current
   1. PID loop
1. The beamline engineers are aiming to test/build the new PSU system by December 2020 (so that it is avail;able before the start of the long shutdown).

**Comment by kjwoodsISIS** [comment]
Reply from Jonny R:

> The Solenoid controller will be an Omron Sysmac NX1 (NX1029000).

**Comment by FreddieAkeroyd** [comment]
Expected for December 2020

**Comment by DominicOram** [comment]
To do this we will need a description of the addresses on the PLC. @Kevin-Woods-Tessella do we have this yet?

**Comment by kjwoodsISIS** [comment]
@DominicOram  - I have asked several times for the PLC memory map (most recently, this morning), but it has yet to be provided.  The RIKEN beamline team know we need it.

**Comment by kjwoodsISIS** [comment]
/remind me to add this issue to a sprint **1st March 2021**

**Comment by DominicOram** [comment]
reminder bot seems to have broken :(

**Comment by DominicOram** [comment]
/remind me in 1 minute if this is still broken

**Comment by kjwoodsISIS** [comment]
Original description (retained for information purposes only):
As a beamline engineer, I need an interface (OPI) to the new PSU for the RIKEN solenoid, so that I can acquire information about the status of the PSU.  Direct control of the PSU will be via a PLC, so IBEX will need to acquire information about the PSU from the PLC.

#### Acceptance Criteria
1. I can view the status of the PSU via the IBEX GUI.
   * Suggested layout of OPI:
   * Main Tab
      1. Current readback
      1. Voltage readback
      1. Ramp Rate setting
   * Advanced Tab (Interlock Status)
      1. Magnet Coil Temperature High (red LED, lit if high)
      1. Cryostat SHE Pressure High (red LED, lit if high)
      1. Powerlead Temperature Emergency (red LED, lit if in emergency alarm)
      1. Powerlead Temperature Emergency High (red LED, lit if high)
      1. Cryostat Vacuum > 10E-03 Torr (red LED, lit if > 10E-03 Torr)
      1. Cryostat Vacuum > 10E-04 Torr (red LED, lit if > 10E-04 Torr)
      1. Refrigerator Trip (red LED, lit if tripped)
      1. Cooldown Ready (green LED, lit if ready)
      1. Excitation Ready (green LED, lit if ready)
      1. Discharge Valve Open (green LED, lit if open)
1. I can obtain the status of the PSU via genie_python scripts.

#### Notes
1. The RIKEN solenoid is an existing device on the RIKEN beamline.  It is an old system and is due to be replaced within the next 2-3 years.
   * Because it is an old system, it has an equally old power supply.  There is a risk that the old power supply might fail before a new solenoid can be installed.
   * The old power supply is not controlled via IBEX, nor is there any intention that it should be (the objective of this ticket is to provide facilities for controlling the new PSU).
   * The beamline engineers would like to have a spare PSU that they can have ready, "sitting-on-the-shelf" to replace the old PSU.  The spare PSU will communicate with IBEX (via a PLC).
1. The spare PSU will be a TDK Lambda Genesys power supply.
1. The PSU itself is expected to be directly controlled by a PLC ([Omron Sysmac NX1 (NX1029000)](https://industrial.omron.co.uk/en/products/nx1)). IBEX will communicate to the PLC.  
   1. The PLC is expected to control:
      * setting of currents & voltages
      * ramping of the current
      * PID loop
   1. The PLC might also be able to provide 
      * current & voltage readbacks
      * interlock statuses.
   1. IBEX is required to display these items of information on the OPI (the final list will be determined nearer the time).
   1. Use the [photograph of the existing (hardware) control panel](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/RIKEN%20FE/RIKEN%20HMI%20Screens/Solenoid_Control_Panel.jpg) as a guide
1. Communications with the PLC will use the FINS protocol
   1. There have been discussions about using the CIP protocol.  However creating a CIP driver is a big job and is outside of the scope of this ticket.
1. The beamline engineers are aiming to test/build the new PSU system by December 2020 (~so that it is available before the start of the long shutdown~).

**Comment by kjwoodsISIS** [comment]
The Mitsubishi PLC does not exist.  Read on  ...
> From: Searle, Rob
> Sent: 20 November 2020 09:04
> To: Woods, Kevin
> Cc: Thomas, Gary; Ranner, Jonny; Lord, James
> Subject: RE: RIKEN Solenoid PSU & PLC
> 
> The Misubishi PLC was replaced by a remote drop of the ABB Cryogenics PLC about 5 years ago. 
> The Cryo PLC is NOT networked, as is a windows based system. We have been advised by ABB that 
> this system must remain off the network as there are safety concerns over windows updates etc 
> affecting our ability to control the system, and IT will not allow the PC to be on the network 
> with windows updates turned off. (this is also still a windows 7 based system – plans to upgrade 
> are on hold due to the possibility of decommissioning the system if the solenoid is replaced)
> 
> The signals that you are looking at replicating are all digital outputs from the PLC that energise 
> relays in the Mitsubishi rack, these relays in turn drive the indications on the front panel and 
> the interlocks to the solenoid PSU (a copy of the interlocks drawing, 
> RIKEN_Solenoid_Interlocks_Mitsubishi_SI-9820-155-13.pdf, is in the RIKEN FE folder 
> on the ICP Discussion site).
> 
> You would have to pick up the digital signals from the relays and take them into another system 
> to capture them. I believe these are either MY2 or MY4 relays, if the latter then there will be 
> spare contacts available, if they are MY2’s then they could be changed for MY4’s to provide the 
> contacts necessary.
> 
> Cheers
> Rob.

**Comment by kjwoodsISIS** [comment]
> From: Ranner, Jonny
> Sent: 25 January 2021 23:06
> To: Woods, Kevin
> Cc: Thomas, Gary; Bowden, Jack
> Subject: RE: RIKEN Solenoid PSU and new Long Shutdown dates
> 
> Hi Kevin,
> 
> We will still aim for early April. We’re not far off completing the design and will be building the racks soon. Once built, we can plug it into the network and start programming and testing.
> 
> Regards,
> Jonny

**Comment by davidkeymer** [comment]
Draught version of PLC memory map: 
[RIKEN_Solenoid_Fields.xlsx](https://github.com/ISISComputingGroup/IBEX/files/7184767/RIKEN_Solenoid_Fields.xlsx)

As of 18/09/2021, the PLC has not yet been connected to the network.

**Comment by davidkeymer** [comment]
Update on 06/03/2023:

> The PSU rack is nearly fully assembled. I am waiting for input from DL this week for support with the quench detection system. After this I plan to do some basic testing of the PLC and PSU. The system will then be ready for integration/planning for integration with IBEX.

**Comment by GRyall** [comment]
Will this ticket become obsolete when the new Riken Solenoid is delivered this year?

**Comment by KathrynBaker** [comment]
Yes it will, so rather than repurpose this ticket, I'll close it.
